persist/fs/commitlog/reader.go

```diff
@@ -110,52 +157,132 @@ func (r *reader) Read() (
 	annotation ts.Annotation,
 	resultErr error,
 ) {
-	entry, err := r.readEntry()
-	if err != nil {
-		resultErr = err
-		return
+	if r.nextIndex == 0 {
+		go r.readLoop()
+		for i, decoderBuf := range r.decoderBufs {
+			go r.decoderLoop(decoderBuf, r.outBufs[i])
+		}
 	}
+	// Data is written into these channels in round-robin fashion, so reading them
+	// one at a time in the same order results in an ordered stream.
+	// TODO: This doesn't eventually deadlock because the reader goroutine will
+	// continue to fill these channels with EOFs until it is closed. We should
+	// probably avoid the deadlock (read from empty channel) scenario in a more
+	// graceful way.
+	rr := <-r.outBufs[r.nextIndex%r.numConc]
```

Comment:
Yeah we should `rr, ok := <- ....` then `if !ok { return nil, ts.Datapoint{}, xtime.Unit(0), ts.Annotation(nil), errUnexpectedEOF }` or something.